File: algorithms/java/105_construct_binary_tree_from_predorder_and_inorder_traversal/Construct.java
public class Contruct {
    
    // 1. using hashmap
    public TreeNode constructTree1(int[] preorder, int[] inorder) {
        if (preorder == null || preorder.length == 0 || inorder == null || inorder.length == 0) {
            return null;
        }
        Map<Integer, Integer> inorderMap = buildMap(inorder);
        return construct1(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1, inorderMap);
    }

    private TreeNode construct1(int[] preorder, int pStart, int pEnd, int[] inorder, int iStart, int iEnd, Map<Integer, Integer> inorderMap) {
        if (pStart > pEnd) {
            return null;
        }
        TreeNode root = new TreeNode(preorder[pStart]);
        int inIndex = inorderMap.get(root.val);
        int leftSize = inIndex - iStart;
        int rightSize = iEnd - inIndex;
        root.left = construct1(preorder, pStart + 1, pStart + leftSize, inorder, iStart, inIndex - 1, inorderMap);
        root.right = construct1(preorder, pEnd - rightSize + 1, pEnd, inorder, inIndex + 1, iEnd, inorderMap);
        return root;
    }

    private Map<Integer, Integer> buildMap(int[] inorder) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        return map;
    } 

    
    // 2. doing both together
    public TreeNode constructTree2(int[] preorder, int[] inorder) {
        if (preorder == null || preorder.length == 0 || inorder == null || inorder.length == 0) {
            return null;
        }
        return construct2(preorder, inorder, Integer.MAX_VALUE);
    }
    private int preIndex = 0;
    private int inIndex = 0;
    private TreeNode construct2(int[] preorder, int[] inorder, int nextInorderValue) {
        // base case: we need to know when the current subtree is finished
        if (inIndex == inorder.length || inorder[inIndex] == nextInorderValue) return null;
        // preorder
        TreeNode root = new TreeNode(preorder[preIndex]);
        preIndex++;
        root.left = construct2(preorder, inorder, root.val);
        // inorder
        inIndex++; // traverse root itself
        root.right = construct2(preorder, inorder, nextInorderValue);
        return root;
    }
}

File: algorithms/cpp/105_pre_in_construct/Construct.cpp
class Construct {
public:
	// recursion
	TreeNode* buildTree_rec(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> inorder_map;
        for (int i = 0; i < inorder.size(); ++i) {
            inorder_map[inorder[i]] = i;
        }
        return construct1(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1, inorder_map);
    }

    TreeNode* construct1(vector<int>& preorder, int p_start, int p_end, vector<int>& inorder, int i_start, int i_end, unordered_map<int, int>& inorder_map) {
        if (p_start > p_end)
            return NULL;
        TreeNode* root = new TreeNode(preorder[p_start]);
        int in_index = inorder_map[root->val];
        int left_size = in_index - i_start;
        int right_size = i_end - in_index;
        root->left = construct1(preorder, p_start + 1, p_start + left_size, inorder, i_start, in_index - 1, inorder_map);
        root->right = construct1(preorder, p_end - right_size + 1, p_end, inorder, in_index + 1, i_end, inorder_map);
        return root;
    }

	// traverse while construct
	TreeNode* buildTree_traverse(vector<int>& preorder, vector<int>& inorder) {
        return construct2(preorder, inorder, INT_MAX);
    }
    int pre_index = 0;
    int in_index = 0;
    TreeNode* construct2(vector<int>& preorder, vector<int>& inorder, int next_inorder_val) {
        if (in_index == inorder.size() || inorder[in_index] == next_inorder_val)
            return NULL;
        TreeNode* root = new TreeNode(preorder[pre_index]);
        ++pre_index;
        root->left = construct2(preorder, inorder, root->val);
        ++in_index;
        root->right = construct2(preorder, inorder, next_inorder_val);
        return root;
    }
};
